src/Action/JsCaptureAction.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Prometee\PayumStripe\Action;
+
+use Payum\Core\Action\ActionInterface;
+use Payum\Core\Bridge\Spl\ArrayObject;
+use Payum\Core\Exception\LogicException;
+use Payum\Core\Exception\RequestNotSupportedException;
+use Payum\Core\GatewayAwareInterface;
+use Payum\Core\GatewayAwareTrait;
+use Payum\Core\Request\Capture;
+use Payum\Core\Request\Sync;
+use Payum\Core\Security\GenericTokenFactoryAwareInterface;
+use Payum\Core\Security\GenericTokenFactoryAwareTrait;
+use Prometee\PayumStripe\Request\Api\Pay;
+use Prometee\PayumStripe\Request\Api\Resource\CreateCustomer;
+use Prometee\PayumStripe\Request\Api\Resource\CreatePaymentIntent;
+use Prometee\PayumStripe\Request\Api\Resource\RetrieveCustomer;
+use Prometee\PayumStripe\Request\Api\Resource\RetrievePaymentMethod;
+
+class JsCaptureAction extends CaptureAction implements ActionInterface, GatewayAwareInterface, GenericTokenFactoryAwareInterface
+{
+    use GatewayAwareTrait;
+    use GenericTokenFactoryAwareTrait;
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param Capture $request
+     */
+    public function execute($request): void
+    {
+        RequestNotSupportedException::assertSupports($this, $request);
+
+        $model = ArrayObject::ensureArrayObject($request->getModel());
+
+        // If we don't have the PaymentIntent object, we need to create it
+        if (false === $model->offsetExists('id')) {
+            // Create another token to allow payment webhooks to use `Notify`
+            $token = $request->getToken();
+            $notifyToken = $this->tokenFactory->createNotifyToken(
+                $token->getGatewayName(),
+                $token->getDetails()
+            );
+            $this->embedNotifyTokenHash($model, $notifyToken);
+
+            // Get or Create the Customer if we have some data in the model
+            if ($model->offsetExists('customer') && !isset($model['customer']['id'])) {
+                if (isset($model['customer']['stripe_id'])) {
+                    $retrieveCustomer = new RetrieveCustomer($model['customer']['stripe_id']);
+                    $this->gateway->execute($retrieveCustomer);
+
+                    $model['customer'] = $retrieveCustomer->getApiResource();
+                } else {
+                    $createCustomer = new CreateCustomer($model->offsetGet('customer'));
+                    $this->gateway->execute($createCustomer);
+
+                    $model['customer'] = $createCustomer->getApiResource();
+                }
+            }
```

Comment:
This could be moved to an Extension or a specific action because it can be used with other gateways.

I don't understand from where the `stripe_id` come from, do you have a specific code elsewhere which handle it ?
Maybe using customer metadata could be better here ?

Reply:
It's used in the Sylius plugin. The user of that plugin needs to add this Trait to the Customer entity:  https://github.com/odiseoteam/SyliusPayumStripeCheckoutSessionPlugin/blob/master/src/Entity/StripeIdTrait.php. And then is used here: https://github.com/odiseoteam/SyliusPayumStripeCheckoutSessionPlugin/blob/master/src/Provider/CustomerDetailsProvider.php#L52.